impl-tinkerpop/src/main/java/org/hawkular/inventory/impl/tinkerpop/RelationshipService.java

```diff
@@ -16,86 +16,219 @@
+                    Direction.BOTH;
+            final Direction d2 = direction == outgoing ? Direction.IN : direction == incoming ? Direction.OUT :
+                    Direction.BOTH;
+            Stream<Edge> edges = StreamSupport.stream(v.getEdges(d1)
+                    .spliterator(), false)
+                    .filter(edge -> name.equals(edge.getLabel())
+                            && targetOrSource.getId().equals(edge
+                            .getVertex(d2).<String>getProperty(Constants.Property.uid.name())));
+
+            return edges.collect(Collectors.toList()).get(0).getId();
+        };
+
+        HawkularPipeline<?, Object> pipe = null;
+        switch (direction) {
+            case outgoing:
+                pipe = source().linkOut(name, incidenceVertex).transform(transformation);
+                break;
+            case incoming:
+                pipe = source().linkIn(name, incidenceVertex).transform(transformation);
+                break;
+            case both:
+                // basically creates a bi-directional relationship
+                pipe = source().linkBoth(name, incidenceVertex).transform(transformation);
+                break;
+        }
+
+        String newId = pipe.cast(String.class).next();
+        return createSingleBrowser(RelationWith.id(newId));
     }
 
     @Override
-    public void delete(String id) throws EntityNotFoundException {
-        //TODO implement
-        throw new UnsupportedOperationException();
+    public Relationships.Single linkWith(Relationships.WellKnown name, Entity targetOrSource) throws
+            RelationNotFoundException {
+        return linkWith(name.name(), targetOrSource);
+    }
+
+    @Override
+    public void update(Relationship relationship) throws RelationNotFoundException {
+        if (null == relationship) {
+            throw new IllegalArgumentException("relationship was null");
+        }
+        if (null == relationship.getId()) {
+            throw new IllegalArgumentException("relationship's ID was null");
+        }
+
+        // check if the source/target vertex of the relationship is on the current position in hawk-pipe. If not use the
+        // `ifThenElse` for returning an empty iterator and fail subsequent querying
+        PipeFunction<Vertex, Boolean> ifFunction = vertex -> {
+            String uid = vertex.getProperty(Constants.Property.uid.name());
+            boolean sourceOk = uid.equals(relationship.getSource().getId());
+            boolean targetOk = uid.equals(relationship.getTarget().getId());
+            boolean retBool = direction == outgoing ? sourceOk : direction ==
+                    incoming ? targetOk : (sourceOk || targetOk);
+            return retBool;
+        };
+        PipeFunction<Vertex, ?> thenFunction = v -> v;
+        PipeFunction<Vertex, ?> elseFunction = v -> Collections.<Vertex>emptyList().iterator();
+        source().ifThenElse(ifFunction, thenFunction, elseFunction);
```

Comment:
> btw. the `RelationshipService.linkWith()` needs to go twice to the "db"
> 1st time: `convert(targetOrSource);` on L118
> 2nd: the switch in the query
> 
> It's because of the signature of those `link{Out|In|Both}` methods that
> expects a vertex as a parameter, so I need to look it up first. Perhaps
> this could be optimized using those overloaded variants with the named step
> as a parameter. But let's keep it simple and iterate over it, wdyt?
> 

+1, let's go with this now and see how it performs.